[Server] FCM PUSH 모듈

ionc635 opened the pull request:
## 👀 이슈

resolve #126

## 📌 개요

Nest.js에 맞게 FCM PUSH 모듈을 만듭니다.

## 👩‍💻 작업 사항

fcm-node 의존성 설치
fcm module 생성
fcm interface 생성
FCM PUSH 기능 구현

## ✅ 참고 사항

serverKey는 추후 동적 모듈 적용할 계획입니다.

ionc635 committed:
chore: fcm-node 의존성 설치

ionc635 committed:
feat: Fcm module 생성

- fcm.module.ts 파일 생성

app.module.ts
- FcmModule 의존성 추가

ionc635 committed:
feat: fcm interface 생성

fcm.interfaces.ts
- Data 타입 설정(title, body)

ionc635 committed:
feat: FCM PUSH 기능 구현

fcm.service.ts
- 안드로이드, ios PUSH 기능 구현
- serverKey는 동적 모듈 적용 계획

jaryapp commented on packages/server/src/fcm/fcm.service.ts:
```diff
@@ -0,0 +1,49 @@
+    });
+  }
+
+  public send_notice(to: string, data: Data) {
```
함수명은 camelCase로 통일 부탁드려요~

ionc635 replied:
네 확인했습니다!

jaryapp reviewed (approved):
스티븐 충림이 첫 PR 환영합니다 🥳🥳

jaryapp commented on packages/server/src/fcm/fcm.service.ts:
```diff
@@ -0,0 +1,49 @@
+  constructor() {
+    // this.serverKey = this.configService.get("serverKey");
+    // this.fcm = new FCM(this.serverKey);
+    this.message = (to: string, data: Data, collapse_key: string) => ({
```
message함수를 class의 멤버함수가 아니라 멤버변수로 정의한 이유가 있을까요?

ionc635 replied:
특별한 이유는 없습니다. 수정하는게 좋을까요?

jaryapp replied:
@ionc635 
넵 함수는 멤버함수로 관리하는것이 통일성 있을것 같아요~

ionc635 replied:
네 확인했습니다!

ionc635 committed:
style: 함수명 스타일 수정

- snake case에서 camel case로 수정

ionc635 committed:
style: Prettier 추가 적용

- fcm.module.ts

ionc635 committed:
refactor: message 변수를 함수로 정의

apply review

JeongHyeongKim commented on packages/server/src/fcm/fcm.service.ts:
```diff
@@ -0,0 +1,51 @@
+    };
+  }
+
+  public sendNotice(to: string, data: Data) {
+    this.fcm.send(
+      this.message(to, data, "com.jaryapp.myapplication3"),
+      (err, res) => {
+        if (err) {
+          console.log(err);
+          console.log("Something has gone wrong!");
+        } else {
+          console.log("Successfully sent with response: ", res);
+        }
+      },
+    );
+  }
```
아래 function의 경우, IOS를 위해 쓰이는것이 명확하나, comment 드린 function은 어떤 체제에서 동작하는지 명확하지 않아 아래 function naming과 통일하는 것이 좋아보입니다.

ex) sendNoticeAndroid

ionc635 replied:
수정했습니다!

JeongHyeongKim commented on packages/server/src/fcm/fcm.service.ts:
```diff
@@ -0,0 +1,51 @@
+    // this.fcm = new FCM(this.serverKey);
+  }
+
+  private message(to: string, data: Data, collapse_key: string) {
+    return {
+      to,
+      collapse_key,
+      notification: {
+        title: data.title,
+        body: data.body,
+        click_action: "Result3Activity",
+      },
+      data,
+    };
+  }
```
title, body가 not nullable한 속성이라면, 위에 검사하는 로직을 넣는게 좋을거같습니다.

```suggestion
  private message(to: string, data: Data, collapse_key: string) {
    const {title, body} = data;
    
    if(_.isNil(title) || _.isNil(body)){
        // some logics...
    }
    return {
      to,
      collapse_key,
      notification: {
        title: data.title,
        body: data.body,
        click_action: "Result3Activity",
      },
      data,
    };
  }
```

ionc635 replied:
수정했습니다.

JeongHyeongKim commented on packages/server/src/fcm/fcm.service.ts:
```diff
@@ -0,0 +1,51 @@
+import { Injectable } from "@nestjs/common";
+import * as FCM from "fcm-node";
+import { Data } from "./fcm.interfaces";
```
path 설정이 잘 안되어있네요
나중에 제가 path 설정 일괄적으로 적용하겠습니다.

ionc635 replied:
확인했습니다!

ionc635 committed:
refactor: sendNotice 함수명 변경

fcm.service.ts
- sendNotice에서 sendNoticeAndroid로 변경

ionc635 committed:
refactor: FCM Push 비즈니스 로직 보완

fcm.service.ts
- message 값(title, body) 없을 시 예외 처리
- 알 수 없는 서버 에러 예외 처리

package.json
- lodash 의존성 설치

jaryapp reviewed (approved):
😁😁
